quil-rs/src/expression/simplification/by_hand.rs

```diff
@@ -160,671 +164,714 @@ fn mul_matches(left_ax: &Expression, right_ax: &Expression) -> bool {
+            }
+
+            // Subtraction "associative" (not really), right: a - (b - c) = (a + c) - b
+            (
+                ref a,
+                InfixOperator::Minus,
+                ref right @ Expression::Infix(InfixExpression {
+                    left: ref b,
+                    operator: InfixOperator::Minus,
+                    right: ref c,
+                }),
+            ) => {
+                let original = sub!(a.clone(), right.clone());
+                let ac = simplify_infix(a, InfixOperator::Plus, c, limit - 1);
+                let new = simplify_infix(&ac, InfixOperator::Minus, b, limit - 1);
+                min_by_key(original, new, size)
+            }
+
+            // Subtraction "associative" (not really), left: (a - b) - c = (a + c) - b
+            (
+                ref left @ Expression::Infix(InfixExpression {
+                    left: ref a,
+                    operator: InfixOperator::Minus,
+                    right: ref b,
+                }),
+                InfixOperator::Minus,
+                ref c,
+            ) => {
+                let original = sub!(left.clone(), c.clone());
+                let ac = simplify_infix(a, InfixOperator::Plus, c, limit - 1);
+                let new = simplify_infix(&ac, InfixOperator::Minus, b, limit - 1);
+                min_by_key(original, new, size)
+            }
+
+            // Division "associative" (not really), right: a / (b / c) = (a * c) / b
+            (
+                ref a,
+                InfixOperator::Slash,
+                ref right @ Expression::Infix(InfixExpression {
+                    left: ref b,
+                    operator: InfixOperator::Slash,
+                    right: ref c,
+                }),
+            ) => {
+                let original = div!(a.clone(), right.clone());
+                let ac = simplify_infix(a, InfixOperator::Star, c, limit - 1);
+                let new = simplify_infix(&ac, InfixOperator::Slash, b, limit - 1);
+                min_by_key(original, new, size)
+            }
+
+            // Division "associative" (not really), left: (a / b) / c = a / (b * c)
+            (
+                ref left @ Expression::Infix(InfixExpression {
+                    left: ref a,
+                    operator: InfixOperator::Slash,
+                    right: ref b,
+                }),
                 InfixOperator::Slash,
-                multiplier,
-                limit.saturating_sub(1),
-            );
-            let new = simplify_infix(
-                &new_multiplier,
+                ref c,
+            ) => {
+                let original = div!(left.clone(), c.clone());
+                let bc = simplify_infix(b, InfixOperator::Star, c, limit - 1);
+                let new = simplify_infix(a, InfixOperator::Slash, &bc, limit - 1);
+                min_by_key(original, new, size)
+            }
+
+            // Right distribution: a * (b + c) = (a * b) + (a * c)
```

Comment:
Could merge with the next case, I think

Reply:
Same comment as before re: reconstructing the original